### PR TITLE
Tcat zephyr implementation and openthread upmerge

### PR DESCRIPTION
**Pull request by canisLupus1313**
This PR introduces new version of openthread `4ed44bc`.

Introduces also implementation of TCAT feature in Zephyr platform.

### CODE REVIEW

File: modules/openthread/platform/ble.c [hunk 1]
```diff
@@ -0,0 +1,508 @@
+/*
+ *  Copyright (c) 2022, The OpenThread Authors.
```
Comment (edmont) [inline modules/openthread/platform/ble.c]:
```suggestion
 *  Copyright (c) 2023, The OpenThread Authors.
```

**Review by edmont — COMMENTED** [review]
It seems an initial compliance cleanup is needed.

**Review by maciejbaczmanski — COMMENTED** [review]
I'd suggest to keep all logs regarding errors as `LOG_WRN` level

File: modules/openthread/platform/ble.c [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+/* Zephyr Kernel Objects */
+
+static void otPlatBleThread(void *, void *, void *);
+uint8_t otPlatBleMsgBuf[PLAT_BLE_MSG_DATA_MAX];
+
+K_SEM_DEFINE(otPlatBleInitSemaphor, 0, 1);
+K_SEM_DEFINE(otPlatBleEventSemaphor, 0, 10000);
+RING_BUF_DECLARE(otPlatBleRingBuf, PLAT_BLE_RING_BUF_SIZE);
+K_THREAD_DEFINE(otPlatBleTid, PLAT_BLE_THREAD_STACK_SIZE, otPlatBleThread, NULL, NULL, NULL, 5, 0,
+		PLAT_BLE_THREAD_DEALY);
```
Comment (rlubos) [inline modules/openthread/platform/ble.c:68]: All those variables could be `static`? And please use `snake_case` for Zephyr code whenever possible (I am aware there have to exceptions for platform functions, but it's not the case for those variables).

File: modules/openthread/platform/ble.c [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+
+static void otPlatBleThread(void *unused1, void *unused2, void *unused3)
+{
+	ARG_UNUSED(unused1);
+	ARG_UNUSED(unused2);
+	ARG_UNUSED(unused3);
+
+	uint8_t len;
+	int8_t rssi;
+	otBleRadioPacket myPacket;
```
Comment (rlubos) [inline modules/openthread/platform/ble.c:155]: Ditto, snake_case plase.

File: modules/openthread/platform/ble.c [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+{
+	otError error = OT_ERROR_NONE;
+
+	if (aLen <= PLAT_BLE_MSG_DATA_MAX && aData == NULL) {
+		return OT_ERROR_INVALID_ARGS;
+	}
+
+	k_sched_lock();
+
+	if (ring_buf_space_get(&otPlatBleRingBuf) >=
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/ble.c]: fixed.

File: modules/openthread/platform/ble.c [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+
+#define PLAT_BLE_MSG_CONNECT    0xFE
+#define PLAT_BLE_MSG_DISCONNECT 0xFF
+
+/* Zephyr Kernel Objects */
+
+static void otPlatBleThread(void *, void *, void *);
+uint8_t otPlatBleMsgBuf[PLAT_BLE_MSG_DATA_MAX];
+
+K_SEM_DEFINE(otPlatBleInitSemaphor, 0, 1);
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/ble.c]: fixed.

File: modules/openthread/platform/ble.c (hunk 3 above)
Comment (canisLupus1313) [reply modules/openthread/platform/ble.c]: fixed.

File: modules/openthread/platform/ble.c (hunk 2 above)
Comment (canisLupus1313) [reply modules/openthread/platform/ble.c]: fixed.

File: modules/openthread/platform/ble.c [hunk 6]
```diff
@@ -0,0 +1,462 @@
...
+	BT_DATA(BT_DATA_NAME_COMPLETE, DEVICE_NAME, DEVICE_NAME_LEN),
+};
+
+static const struct bt_data sd[] = {
+	BT_DATA_BYTES(BT_DATA_UUID16_ALL, BT_UUID_16_ENCODE(MY_SERVICE_UUID)),
+};
+
+/* Zephyr BLE Message Queue and Thread */
+
+static bool otPlatBleQueueMsg(const uint8_t *aData, uint8_t aLen, int8_t aRssi)
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/ble.c:118]: Yes it's not yet implemented.

File: modules/openthread/platform/ble.c [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+
+otError otPlatBleEnable(otInstance *aInstance)
+{
+	int err;
+
+	otPlatBleOpenThreadInstance = aInstance;
+	err = bt_enable(bt_ready);
+
+	if (err != 0 && err != -EALREADY) {
+		LOG_INF("BLE enable failed with error code %d", err);
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/ble.c]: fixed.

File: modules/openthread/platform/ble.c [hunk 8]
```diff
@@ -0,0 +1,462 @@
...
+		}
+
+		LOG_INF("Connection parameters updated (mtu=%" PRIu16 ")!", mtu);
+	}
+}
+
+static void bt_ready(int err)
+{
+	if (err) {
+		LOG_INF("BLE init failed with error code %d", err);
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/ble.c]: fixed.

File: modules/openthread/platform/ble.c [hunk 9]
```diff
@@ -0,0 +1,462 @@
...
+otError otPlatBleGapAdvStart(otInstance *aInstance, uint16_t aInterval)
+{
+	ARG_UNUSED(aInstance);
+	ARG_UNUSED(aInterval);
+
+	/* TO DO advertisement format change */
+	int err = bt_le_adv_start(BT_LE_ADV_CONN, ad, ARRAY_SIZE(ad), sd, ARRAY_SIZE(sd));
+
+	if (err != 0 && err != -EALREADY) {
+		LOG_INF("Advertising failed to start (err %d)", err);
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/ble.c]: fixed.

File: modules/openthread/platform/ble.c [hunk 10]
```diff
@@ -0,0 +1,462 @@
...
+{
+	struct bt_conn_info info;
+	char addr[BT_ADDR_LE_STR_LEN];
+	uint16_t mtu;
+	otError error = OT_ERROR_NONE;
+
+	otPlatBleConnection = bt_conn_ref(conn);
+
+	if (err) {
+		LOG_INF("Connection failed (err %u)", err);
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/ble.c]: fixed.

File: modules/openthread/platform/ble.c [hunk 11]
```diff
@@ -0,0 +1,462 @@
...
+	}
+
+	if (otPlatBleConnection == NULL) {
+		return OT_ERROR_INVALID_STATE;
+	}
+
+	/* Check whether notifications are enabled or not */
+	if (bt_gatt_is_subscribed(otPlatBleConnection, attr, BT_GATT_CCC_NOTIFY)) {
+		if (bt_gatt_notify_cb(otPlatBleConnection, &params)) {
+			LOG_INF("Error, unable to send notification");
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/ble.c]: fixed.

File: modules/openthread/platform/ble.c [hunk 12]
```diff
@@ -0,0 +1,462 @@
...
+			return OT_ERROR_INVALID_ARGS;
+		}
+	} else {
+		LOG_INF("Warning, notification not enabled on the selected attribute");
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/ble.c]: fixed.

File: modules/openthread/platform/ble.c [hunk 13]
```diff
@@ -0,0 +1,462 @@
...
+	uint16_t mtu;
+	otError error = OT_ERROR_NONE;
+
+	otPlatBleConnection = bt_conn_ref(conn);
+
+	if (err) {
+		LOG_INF("Connection failed (err %u)", err);
+		return;
+	} else if (bt_conn_get_info(conn, &info)) {
+		LOG_INF("Could not parse connection info");
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/ble.c]: fixed.

File: modules/openthread/platform/ble.c [hunk 14]
```diff
@@ -0,0 +1,462 @@
...
+
+	const struct bt_gatt_attr *attr = &my_service.attrs[3];
+
+	struct bt_gatt_notify_params params = {.uuid = BT_UUID_TCAT_SERVICE_TX,
+					       .attr = attr,
+					       .data = aPacket->mValue,
+					       .len = aPacket->mLength,
+					       .func = on_sent};
+
+	LOG_INF("Send data, handle %d, len %d", attr->handle, aPacket->mLength);
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/ble.c]: fixed.

File: modules/openthread/platform/ble.c [hunk 15]
```diff
@@ -0,0 +1,462 @@
...
+		LOG_WRN("Error queuing message: %s", otThreadErrorToString(error));
+	}
+
+	return len;
+}
+
+/* This function is called whenever a Notification has been sent by the TX Characteristic */
+static void on_sent(struct bt_conn *conn, void *user_data)
+{
+	ARG_UNUSED(user_data);
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/ble.c:208]: fixed.

File: modules/openthread/platform/ble.c [hunk 16]
```diff
@@ -0,0 +1,462 @@
...
+		}
+
+		LOG_INF("CCCD update (mtu=%" PRIu16 ")!", mtu);
+
+		break;
+
+	case BT_GATT_CCC_INDICATE:
+		break;
+
+	case 0:
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/ble.c]: fixed.

File: modules/openthread/platform/ble.c [hunk 17]
```diff
@@ -0,0 +1,459 @@
...
+
+	uint16_t len;
+	int8_t rssi;
+	otBleRadioPacket my_packet;
+
+	LOG_INF("%s started", __func__);
+
+	while (1) {
+		k_sem_take(&ot_plat_ble_event_semaphore, K_FOREVER);
+		ring_buf_get(&ot_plat_ble_ring_buf, (uint8_t*)&len, sizeof(len));
```
Comment (maciejbaczmanski) [inline modules/openthread/platform/ble.c:161]:
```suggestion
		ring_buf_get(&ot_plat_ble_ring_buf, (uint16_t*)&len, sizeof(len));
```

File: modules/openthread/platform/ble.c [hunk 18]
```diff
@@ -0,0 +1,459 @@
...
+	if (aLen <= PLAT_BLE_MSG_DATA_MAX && aData == NULL) {
+		return OT_ERROR_INVALID_ARGS;
+	}
+
+	k_sched_lock();
+
+	len = sizeof(aLen) + sizeof(aRssi) + ((aLen <= PLAT_BLE_MSG_DATA_MAX) ? aLen : 0);
+
+	if (ring_buf_space_get(&ot_plat_ble_ring_buf) >= len) {
+		ring_buf_put(&ot_plat_ble_ring_buf, (uint8_t*)&aLen, sizeof(aLen));
```
Comment (maciejbaczmanski) [inline modules/openthread/platform/ble.c:132]:
```suggestion
		ring_buf_put(&ot_plat_ble_ring_buf, (uint16_t *)&aLen, sizeof(aLen));
```

**Review by maciejbaczmanski — COMMENTED** [review]
missed one more type cast

File: modules/openthread/platform/ble.c (hunk 18 above)
Comment (maciejbaczmanski) [reply modules/openthread/platform/ble.c:132]: nvm,type cast is ok, my bad

Comment (canisLupus1313) [reply modules/openthread/platform/ble.c:132]: This is to silence the compiler warning since `ring_buf_get` takes `const uint8_t*` I have to cast `len`.

**Review by maciejbaczmanski — DISMISSED** [review]

File: modules/openthread/platform/ble.c [hunk 19]
```diff
@@ -0,0 +1,462 @@
...
+
+#define LOG_MODULE_NAME net_openthread_tcat
+#define LOG_LEVEL       CONFIG_OPENTHREAD_LOG_LEVEL
+
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#define DEVICE_NAME     CONFIG_BT_DEVICE_NAME
+#define DEVICE_NAME_LEN (sizeof(DEVICE_NAME) - 1)
+
+#define TOBLE_SERVICE_UUID 0xfffb
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/ble.c:39]: Added a comment that those values are taken from thread speck.

File: modules/openthread/platform/ble.c [hunk 20]
```diff
@@ -0,0 +1,459 @@
...
+#define PLAT_BLE_MSG_CONNECT    (PLAT_BLE_MSG_DATA_MAX + 1U)
+#define PLAT_BLE_MSG_DISCONNECT (PLAT_BLE_MSG_CONNECT + 1U)
+
+/* Zephyr Kernel Objects */
+
+static void ot_plat_ble_thread(void *, void *, void *);
+static uint8_t ot_plat_ble_msg_buf[PLAT_BLE_MSG_DATA_MAX];
+
+static K_SEM_DEFINE(ot_plat_ble_init_semaphore, 0, 1);
+static K_SEM_DEFINE(ot_plat_ble_event_semaphore, 0, 10000);
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/ble.c]: fixed.

File: modules/openthread/platform/ble.c [hunk 21]
```diff
@@ -0,0 +1,459 @@
...
+		} else if (len == PLAT_BLE_MSG_CONNECT) {
+			otPlatBleGapOnConnected(ble_openthread_instance, 0);
+		} else if (len == PLAT_BLE_MSG_DISCONNECT) {
+			otPlatBleGapOnDisconnected(ble_openthread_instance, 0);
+		}
+		openthread_api_mutex_unlock(openthread_get_default_context());
+	}
+}
+
+/*Zephyr BLE service callbacks */
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply modules/openthread/platform/ble.c]: fixed.

**Review by maciejbaczmanski — DISMISSED** [review]

**Review by edmont — DISMISSED** [review]

**Review by maciejbaczmanski — APPROVED** [review]

**Review by pjasinski990 — APPROVED** [review]

**Review by edmont — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Pull request merged**